internal/ui/ui_glfw_darwin.go

```diff
@@ -17,224 +17,108 @@
-//     break;
-//   }
-//   [cursor push];
-// }
-//
-// static void currentMouseLocation(int* x, int* y) {
-//   NSPoint location = [NSEvent mouseLocation];
-//   *x = (int)(location.x);
-//   *y = (int)(location.y);
-// }
-//
-// static void setAllowFullscreen(uintptr_t windowPtr, bool allowFullscreen) {
-//   NSWindow* window = (NSWindow*)windowPtr;
-//   if (allowFullscreen) {
-//     window.collectionBehavior |= NSWindowCollectionBehaviorFullScreenPrimary;
-//   } else {
-//     window.collectionBehavior &= ~NSWindowCollectionBehaviorFullScreenPrimary;
-//   }
-// }
-import "C"
-
 import (
 	"fmt"
+	"unsafe"
+
+	"github.com/ebitengine/purego/objc"
 
+	"github.com/hajimehoshi/ebiten/v2/internal/cocoa"
 	"github.com/hajimehoshi/ebiten/v2/internal/glfw"
 	"github.com/hajimehoshi/ebiten/v2/internal/graphicsdriver"
 	"github.com/hajimehoshi/ebiten/v2/internal/graphicsdriver/metal"
 	"github.com/hajimehoshi/ebiten/v2/internal/graphicsdriver/opengl"
 )
 
+var class_EbitengineWindowDelegate objc.Class
+
+func init() {
+	class_EbitengineWindowDelegate = objc.AllocateClassPair(objc.GetClass("NSObject"), "EbitengineWindowDelegate", 0)
+	protocol := objc.GetProtocol("NSWindowDelegate")
+	class_EbitengineWindowDelegate.AddProtocol(protocol)
+	// TODO: remove these Ivars and place them in Go since there is only ever one instance of this class
+	class_EbitengineWindowDelegate.AddIvar("origDelegate", objc.ID(0), "@")
+	class_EbitengineWindowDelegate.AddIvar("origResizable", false, "B")
+	origDelegateOffset := class_EbitengineWindowDelegate.InstanceVariable("origDelegate").Offset()
+	origResizableOffset := class_EbitengineWindowDelegate.InstanceVariable("origResizable").Offset()
+	getOrigDelegate := func(self objc.ID) objc.ID {
+		selfPtr := *(**uintptr)(unsafe.Pointer(&self))
+		return *(*objc.ID)(unsafe.Pointer(uintptr(unsafe.Pointer(selfPtr)) + origDelegateOffset))
+	}
+	getResizable := func(self objc.ID) bool {
+		selfPtr := *(**uintptr)(unsafe.Pointer(&self))
+		return *(*bool)(unsafe.Pointer(uintptr(unsafe.Pointer(selfPtr)) + origResizableOffset))
+	}
+	setResizable := func(self objc.ID, resizable bool) {
+		selfPtr := *(**uintptr)(unsafe.Pointer(&self))
+		*(*bool)(unsafe.Pointer(uintptr(unsafe.Pointer(selfPtr)) + origResizableOffset)) = resizable
+	}
+	pushResizableState := func(self, w objc.ID) {
+		window := cocoa.NSWindow{ID: w}
+		setResizable(self, window.StyleMask()&cocoa.NSWindowStyleMaskResizable != 0)
+		if !getResizable(self) {
+			window.SetStyleMask(window.StyleMask() | cocoa.NSWindowStyleMaskResizable)
+		}
+	}
+	popResizableState := func(self, w objc.ID) {
+		window := cocoa.NSWindow{ID: w}
+		if !getResizable(self) {
+			window.SetStyleMask(window.StyleMask() & ^uint(cocoa.NSWindowStyleMaskResizable))
+		}
+		setResizable(self, false)
+	}
+	class_EbitengineWindowDelegate.AddMethod(objc.RegisterName("initWithOrigDelegate:"), objc.NewIMP(func(self objc.ID, _cmd objc.SEL, origDelegate objc.ID) objc.ID {
```

Comment:
Do we need `_` for `cmd`?

Reply:
Well it's convention but I can remove it

Reply:
Let's use `cmd` then. (I'll update the gamepad side)